Issue with Concurrent API Deployment in Lazy-Loading Tenant APIs (On-Demand Loading Enabled)

### Description

Hi Team,

The following exception with an API in a tenant. After the initial failure, they continuously observed 404 errors for this API. This is a tenant API (lazy-loading tenant).

```toml
[apim.sync_runtime_artifacts.gateway]  
gateway_labels = ["Default"]  
enable_on_demand_loading = true  
```

Exception:

```
[2025-09-23 12:10:52,752] ERROR - RestApiAdmin Error trying to add the API to the ESB configuration : PizzaShackAPI:v1.0.0 :: Duplicate resource definition by the name: PizzaShackAPI:v1.0.0
org.apache.synapse.SynapseException: Duplicate resource definition by the name: PizzaShackAPI:v1.0.0
	at org.apache.synapse.config.SynapseConfiguration.handleException(SynapseConfiguration.java:1805) ~[synapse-core_4.0.0.wso2v20_70.jar:4.0.0-wso2v20.70]
	at org.apache.synapse.config.SynapseConfiguration.addAPI(SynapseConfiguration.java:461) ~[synapse-core_4.0.0.wso2v20_70.jar:4.0.0-wso2v20.70]
	at org.apache.synapse.config.SynapseConfiguration.addAPI(SynapseConfiguration.java:441) ~[synapse-core_4.0.0.wso2v20_70.jar:4.0.0-wso2v20.70]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApi(RestApiAdmin.java:858) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApiFromString(RestApiAdmin.java:113) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApiForTenant(RestApiAdmin.java:139) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.apimgt.gateway.utils.RESTAPIAdminServiceProxy.addApi_aroundBody0(RESTAPIAdminServiceProxy.java:52) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.utils.RESTAPIAdminServiceProxy.addApi(RESTAPIAdminServiceProxy.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.service.APIGatewayAdmin.deployAPI_aroundBody96(APIGatewayAdmin.java:760) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.service.APIGatewayAdmin.deployAPI(APIGatewayAdmin.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer.deployAllAPIsAtGatewayStartup_aroundBody6(InMemoryAPIDeployer.java:233) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer.deployAllAPIsAtGatewayStartup(InMemoryAPIDeployer.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsAtStartup_aroundBody2(GatewayStartupListener.java:143) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsAtStartup(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsInGateway_aroundBody20(GatewayStartupListener.java:312) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsInGateway(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.access$0_aroundBody32(GatewayStartupListener.java:298) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.access$0(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener$AsyncAPIDeployment.run_aroundBody0(GatewayStartupListener.java:432) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener$AsyncAPIDeployment.run(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
[2025-09-23 12:10:52,755] ERROR - InMemoryAPIDeployer Error in deploying PizzaShackAPI to the Gateway 
org.apache.axis2.AxisFault: Error while publishing API to the Gateway. Error trying to add the API to the ESB configuration : PizzaShackAPI:v1.0.0 :: Duplicate resource definition by the name: PizzaShackAPI:v1.0.0
	at org.wso2.carbon.apimgt.gateway.utils.RESTAPIAdminServiceProxy.addApi_aroundBody0(RESTAPIAdminServiceProxy.java:55) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.utils.RESTAPIAdminServiceProxy.addApi(RESTAPIAdminServiceProxy.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.service.APIGatewayAdmin.deployAPI_aroundBody96(APIGatewayAdmin.java:760) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.service.APIGatewayAdmin.deployAPI(APIGatewayAdmin.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer.deployAllAPIsAtGatewayStartup_aroundBody6(InMemoryAPIDeployer.java:233) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer.deployAllAPIsAtGatewayStartup(InMemoryAPIDeployer.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsAtStartup_aroundBody2(GatewayStartupListener.java:143) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsAtStartup(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsInGateway_aroundBody20(GatewayStartupListener.java:312) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.deployArtifactsInGateway(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.access$0_aroundBody32(GatewayStartupListener.java:298) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener.access$0(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener$AsyncAPIDeployment.run_aroundBody0(GatewayStartupListener.java:432) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at org.wso2.carbon.apimgt.gateway.listeners.GatewayStartupListener$AsyncAPIDeployment.run(GatewayStartupListener.java:1) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.wso2.carbon.rest.api.APIException: Error trying to add the API to the ESB configuration : PizzaShackAPI:v1.0.0 :: Duplicate resource definition by the name: PizzaShackAPI:v1.0.0
	at org.wso2.carbon.rest.api.service.RestApiAdmin.handleException(RestApiAdmin.java:919) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApi(RestApiAdmin.java:900) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApiFromString(RestApiAdmin.java:113) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApiForTenant(RestApiAdmin.java:139) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.apimgt.gateway.utils.RESTAPIAdminServiceProxy.addApi_aroundBody0(RESTAPIAdminServiceProxy.java:52) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	... 14 more
Caused by: org.apache.synapse.SynapseException: Duplicate resource definition by the name: PizzaShackAPI:v1.0.0
	at org.apache.synapse.config.SynapseConfiguration.handleException(SynapseConfiguration.java:1805) ~[synapse-core_4.0.0.wso2v20_70.jar:4.0.0-wso2v20.70]
	at org.apache.synapse.config.SynapseConfiguration.addAPI(SynapseConfiguration.java:461) ~[synapse-core_4.0.0.wso2v20_70.jar:4.0.0-wso2v20.70]
	at org.apache.synapse.config.SynapseConfiguration.addAPI(SynapseConfiguration.java:441) ~[synapse-core_4.0.0.wso2v20_70.jar:4.0.0-wso2v20.70]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApi(RestApiAdmin.java:858) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApiFromString(RestApiAdmin.java:113) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.rest.api.service.RestApiAdmin.addApiForTenant(RestApiAdmin.java:139) ~[org.wso2.carbon.rest.api_4.7.175.14.jar:?]
	at org.wso2.carbon.apimgt.gateway.utils.RESTAPIAdminServiceProxy.addApi_aroundBody0(RESTAPIAdminServiceProxy.java:52) ~[org.wso2.carbon.apimgt.gateway_9.28.116.279.jar:?]
	... 14 more

```
We were able to reproduce the issue in debug mode and identify the root cause as follows:

### Analysis and Root Cause

When a tenant API that is not yet loaded into the GW memory receives a request, it triggers the tenant loading process, which begins loading all APIs under that tenant.

At the same time, the request is dispatched to the **DefaultAPIHandler**\[1]. There, the GW checks whether on-demand loading is enabled (`enable_on_demand_loading`) and, if so, whether the Synapse artifact of the API is already deployed. If not, the passthrough thread attempts to deploy the Synapse artifact for that API only. Meanwhile, the tenant loading implementation\[3] is loading all APIs of the tenant.

This results in two concurrent threads trying to deploy the same API:

* **Thread 1** – Triggered by tenant loading, responsible for deploying all APIs in the tenant.
* **Thread 2** – The passthrough thread, which dispatches the request and tries to deploy the Synapse artifact of the requested API due to on-demand loading.

### Problem Scenario

1. **Thread 2** starts deploying the Synapse artifact of the API via the `deployAPI`\[4] function. It successfully deploys all dependent artifacts (such as endpoints and sequences). However, before it deploys the actual API Synapse artifact at \[4], **Thread 1** begins deploying the same artifact as part of tenant loading.
2. **Thread 1** invokes the `undeployAPI`\[5] function, but since the Synapse API artifact is not yet added, it continues.
3. **Thread 1** then proceeds to undeploy the dependent artifacts, including those deployed by **Thread 2**.
4. While this is happening, **Thread 2** deploys the Synapse artifact of the API. Immediately after, **Thread 1** also tries to deploy the same artifact.
5. This leads to the conflicting behavior the mentioned expection. 


### Steps to Reproduce


1. Create a tenant API and enable the following configuration:

   ```toml
   [apim.sync_runtime_artifacts.gateway]  
   enable_on_demand_loading = true  
   ```

2. Add debug points at:

   * `deployAPI` function \[4] (at the point where we tring call the addAPI restapiAdminServiceProxy.addApi()) 
   * `handleRequestInFlow` function \[2]
     *(Ensure debugging is done in thread mode to avoid freezing the entire JVM and all threads.)*

3. Start the API server. Once the server is up, invoke the API.

   * This will trigger the tenant loading process.
   * The debug point at `handleRequestInFlow` \[2] will be hit.

4. At the same time, the debug point at `deployAPI` \[4] in the tenant loading thread will also be hit.

   * This thread will attempt to load the JWKs endpoint first.
   * Step through the process until it reaches the point of deploying the API you created.
   * Hold the debug point in `deployAPI` \[4] for the tenant loading thread at this stage for your API.

5. Switch to the passthrough thread that was blocked at `handleRequestInFlow` \[2].

   * Step through using *Step Into* until it reaches the `deployAPI` function.
   * Continue stepping until it passes the `unDeployAPI(...)` \[5] point inside `deployAPI`.

6. Switch back to the tenant loading thread.

   * Resume (skip) the debug point so the thread can proceed and add the artifact into the GW.

7. Finally, resume (skip) the passthrough thread as well.

[1] https://github.com/wso2/carbon-apimgt/blob/7d665c0f22cc5e0ca0c4301a2f9d1da5445eaa41/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/DefaultAPIHandler.java#L63
[2] https://github.com/wso2/carbon-apimgt/blob/7d665c0f22cc5e0ca0c4301a2f9d1da5445eaa41/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/DefaultAPIHandler.java#L66
[3] https://github.com/wso2/carbon-apimgt/blob/7d665c0f22cc5e0ca0c4301a2f9d1da5445eaa41/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayStartupListener.java#L423
[4] https://github.com/wso2/carbon-apimgt/blob/7d665c0f22cc5e0ca0c4301a2f9d1da5445eaa41/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APIGatewayAdmin.java#L759
[5] https://github.com/wso2/carbon-apimgt/blob/7d665c0f22cc5e0ca0c4301a2f9d1da5445eaa41/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/service/APIGatewayAdmin.java#L667


### Version

4.2.0

### Environment Details (with versions)

_No response_

--- a/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/APIMgtGatewayConstants.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/APIMgtGatewayConstants.java
@@ -124,6 +124,14 @@ public class APIMgtGatewayConstants {
     public static final String GATEWAY_TYPE = "SYNAPSE";
 
     /**
+     * Lock key prefix used by APILockManager to coordinate concurrent on-demand API deployments.
+     * Appended with the API context to form a unique per-API key.
+     *
+     * @see org.wso2.carbon.apimgt.gateway.utils.APILockManager
+     */
+    public static final String API_LOADING_ON_DEMAND = "LoadAPI_OnDemand";
+
+    /**
      * Constants for Open Tracing
      */
     public static final String SERVICE_NAME = "API:Gateway";

--- a/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/InMemoryAPIDeployer.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/InMemoryAPIDeployer.java
@@ -47,6 +47,7 @@ import org.wso2.carbon.apimgt.gateway.internal.DataHolder;
 import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
 import org.wso2.carbon.apimgt.gateway.service.APIGatewayAdmin;
 import org.wso2.carbon.apimgt.gateway.notifiers.DeploymentStatusNotifier;
+import org.wso2.carbon.apimgt.gateway.utils.APILockManager;
 import org.wso2.carbon.apimgt.impl.APIConstants;
 import org.wso2.carbon.apimgt.impl.APIManagerConfiguration;
 import org.wso2.carbon.apimgt.impl.caching.CacheInvalidationServiceImpl;
@@ -71,6 +72,8 @@ import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
 import java.util.UUID;
+
+import static org.wso2.carbon.apimgt.gateway.APIMgtGatewayConstants.API_LOADING_ON_DEMAND;
 
 /**
  * This class contains the methods used to retrieve artifacts from a storage and deploy and undeploy the API in gateway.
@@ -276,8 +279,42 @@ public class InMemoryAPIDeployer {
                                 if (redeployChangedAPIs && apiMap != null) {
                                     reDeployAPIs(gatewayAPIDTO, apiMap, assignedGatewayLabels, tenantDomain, apiGatewayAdmin);
                                 } else {
-                                    deployAPIFromDTO(gatewayAPIDTO, apiGatewayAdmin);
-                                    syncAPIPropertiesAcrossComponents(gatewayAPIDTO);
+                                    // Startup / tenant-loading path (redeployChangedAPIs=false).
+                                    // Coordinate with the passthrough on-demand thread via APILockManager:
+                                    // - skip if the API is already deployed or the passthrough thread is deploying it
+                                    // - use tryLockNow (non-blocking) so this thread does not stall startup
+                                    String startupKey = API_LOADING_ON_DEMAND.concat(gatewayAPIDTO.getApiContext());
+                                    if (DataHolder.getInstance().isDeployed(gatewayAPIDTO)
+                                            || APILockManager.getInstance().isLocked(startupKey)) {
+                                        if (log.isDebugEnabled()) {
+                                            log.debug("API " + gatewayAPIDTO.getName()
+                                                    + " is already deployed or on-demand deployment is in progress."
+                                                    + " Skipping tenant-startup deploy.");
+                                        }
+                                        continue;
+                                    }
+                                    if (APILockManager.getInstance().tryLockNow(startupKey)) {
+                                        try {
+                                            deployAPIFromDTO(gatewayAPIDTO, apiGatewayAdmin);
+                                            syncAPIPropertiesAcrossComponents(gatewayAPIDTO);
+                                        } finally {
+                                            APILockManager.getInstance().unlock(startupKey);
+                                        }
+                                    } else {
+                                        boolean alreadyDeployed = DataHolder.getInstance().isDeployed(gatewayAPIDTO);
+                                        if (log.isDebugEnabled()) {
+                                            if (alreadyDeployed) {
+                                                log.debug("API " + gatewayAPIDTO.getName()
+                                                        + " was deployed by another thread before acquiring the "
+                                                        + "startup lock. Skipping tenant-startup deploy.");
+                                            } else {
+                                                log.debug("Could not acquire startup lock for API "
+                                                        + gatewayAPIDTO.getName()
+                                                        + ". Another thread may be deploying it. "
+                                                        + "Skipping tenant-startup deploy for now.");
+                                            }
+                                        }
+                                    }
                                 }
                             }
                         } catch (AxisFault axisFault) {
@@ -344,9 +381,25 @@ public class InMemoryAPIDeployer {
                                                     api.getName(), api.getVersion(),
                                                     api.getApiProvider(), api.getApiType(),
                                                     api.getContext());
-                unDeployAPI(deployAPIInGatewayEvent);
-                deployAPIFromDTO(gatewayAPIDTO, apiGatewayAdmin);
-                syncAPIPropertiesAcrossComponents(gatewayAPIDTO);
+                // Consideration 2: guard the JMS-reconnect redeployment path as well.
+                // If the passthrough (on-demand) thread is currently holding the lock for
+                // this API context it will deploy the latest revision, so we can safely skip.
+                String redeployKey = API_LOADING_ON_DEMAND.concat(gatewayAPIDTO.getApiContext());
+                if (APILockManager.getInstance().tryLockNow(redeployKey)) {
+                    try {
+                        unDeployAPI(deployAPIInGatewayEvent);
+                        deployAPIFromDTO(gatewayAPIDTO, apiGatewayAdmin);
+                        syncAPIPropertiesAcrossComponents(gatewayAPIDTO);
+                    } finally {
+                        APILockManager.getInstance().unlock(redeployKey);
+                    }
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("On-demand deployment is in progress for API "
+                                + gatewayAPIDTO.getName()
+                                + ". Skipping JMS-reconnect redeploy.");
+                    }
+                }
             } else if (DataHolder.getInstance().getGatewayRegistrationResponse()
                     != APIConstants.GatewayNotification.GatewayRegistrationResponse.ACKNOWLEDGED) {
                 // If the gateway is not registered yet or if it is registered during

--- a/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/DefaultAPIHandler.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/DefaultAPIHandler.java
@@ -31,6 +31,7 @@ import org.wso2.carbon.apimgt.common.gateway.constants.JWTConstants;
 import org.wso2.carbon.apimgt.gateway.APIMgtGatewayConstants;
 import org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer;
 import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
+import org.wso2.carbon.apimgt.gateway.utils.APILockManager;
 import org.wso2.carbon.apimgt.gateway.utils.GatewayUtils;
 import org.wso2.carbon.apimgt.impl.APIConstants;
 import org.wso2.carbon.apimgt.impl.gatewayartifactsynchronizer.exception.ArtifactSynchronizerException;
@@ -38,6 +39,8 @@ import org.wso2.carbon.apimgt.keymgt.model.entity.API;
 import org.wso2.carbon.inbound.endpoint.protocol.websocket.InboundWebsocketConstants;
 
 import java.util.TreeMap;
+
+import static org.wso2.carbon.apimgt.gateway.APIMgtGatewayConstants.API_LOADING_ON_DEMAND;
 
 /**
  * Default API Handler to handle Default Version.
@@ -92,16 +95,24 @@ public class DefaultAPIHandler extends AbstractSynapseHandler {
                 messageContext.setProperty(APIMgtGatewayConstants.API_OBJECT, selectedAPI);
                 if (GatewayUtils.isOnDemandLoading()) {
                     if (!selectedAPI.isDeployed()) {
-                        synchronized ("LoadAPI_".concat(selectedAPI.getContext()).intern()) {
-                            if(!selectedAPI.isDeployed()) {
+                        // APILockManager provides per-API-context mutual exclusion for both
+                        // passthrough threads (blocking lock) and the tenant-loading / JMS thread
+                        // (non-blocking tryLockNow). The outer isDeployed() check is a volatile
+                        // fast-path to skip lock acquisition once the API is already live.
+                        String key = API_LOADING_ON_DEMAND.concat(selectedAPI.getContext());
+                        APILockManager.getInstance().lock(key);
+                        try {
+                            // Re-check: tenant-loading thread may have deployed while we waited
+                            if (!selectedAPI.isDeployed()) {
                                 InMemoryAPIDeployer inMemoryAPIDeployer = new InMemoryAPIDeployer();
-                                try {
-                                    inMemoryAPIDeployer.deployAPI(selectedAPI.getUuid());
-                                } catch (ArtifactSynchronizerException e) {
-                                    log.error("Error while retrieve and deploy artifact for API : " + selectedAPI.getApiId(), e);
-                                    return false;
-                                }
+                                inMemoryAPIDeployer.deployAPI(selectedAPI.getUuid());
                             }
+                        } catch (ArtifactSynchronizerException e) {
+                            log.error("Error while retrieve and deploy artifact for API : "
+                                    + selectedAPI.getApiId(), e);
+                            return false;
+                        } finally {
+                            APILockManager.getInstance().unlock(key);
                         }
                     }
                 }

--- a/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/internal/DataHolder.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/internal/DataHolder.java
@@ -58,7 +58,7 @@ public class DataHolder {
     private Map<String, String> googleAnalyticsConfigMap = new HashMap<>();
     private Map<String, GraphQLSchemaDTO> apiToGraphQLSchemaDTOMap = new HashMap<>();
     private Map<String, List<String>> apiToKeyManagersMap = new HashMap<>();
-    private Map<String,Map<String, API>> tenantAPIMap  = new HashMap<>();
+    private Map<String, Map<String, API>> tenantAPIMap = new ConcurrentHashMap<>();
     private Map<String, Boolean> tenantDeployStatus = new HashMap<>();
     private Map<String, LLMProviderInfo> llmProviderMap = new HashMap<>();
     private final Map<String, APIKeyInfo> apiKeyInfoHashMap = new ConcurrentHashMap<>();
@@ -308,12 +308,8 @@ public class DataHolder {
         if (index != -1) {
             defaultContext = context.substring(0, index);
         }
-        Map<String, API> apiMap;
-        if (tenantAPIMap.containsKey(api.getOrganization())) {
-            apiMap = tenantAPIMap.get(api.getOrganization());
-        } else {
-            apiMap = new HashMap<>();
-        }
+        Map<String, API> apiMap = tenantAPIMap.computeIfAbsent(api.getOrganization(),
+                k -> new ConcurrentHashMap<>());
         API oldAPI = apiMap.get(api.getContext());
         if (oldAPI != null) {
             apiMap.remove(api.getContext());
@@ -325,7 +321,6 @@ public class DataHolder {
         if (api.isDefaultVersion()) {
             apiMap.put(defaultContext, api);
         }
-        tenantAPIMap.put(api.getOrganization(), apiMap);
     }
 
     public void markAPIAsDeployed(GatewayAPIDTO gatewayAPIDTO) {
@@ -340,6 +335,26 @@ public class DataHolder {
                 api.setRevisionId(gatewayAPIDTO.getRevision());
             }
         }
+    }
+
+    /**
+     * Checks whether the API represented by the given DTO is already marked as deployed
+     * in the in-memory tenant API map.
+     *
+     * <p>Used by the tenant-loading thread to skip deploying an API that the passthrough
+     * (on-demand) thread has already fully deployed.
+     *
+     * @param gatewayAPIDTO the DTO identifying the API by tenant domain and API context
+     * @return {@code true} if the API exists in the map and its {@code deployed} flag is
+     *         {@code true}; {@code false} otherwise
+     */
+    public boolean isDeployed(GatewayAPIDTO gatewayAPIDTO) {
+        Map<String, API> apiMap = tenantAPIMap.get(gatewayAPIDTO.getTenantDomain());
+        if (apiMap != null) {
+            API api = apiMap.get(gatewayAPIDTO.getApiContext());
+            return api != null && api.isDeployed();
+        }
+        return false;
     }
 
     public boolean isDuplicateEvent(String tenantDomain, String apiContext, String lastUpdatedEventId) {

--- a/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/APILockManager.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/APILockManager.java
@@ -1,0 +1,149 @@
+/*
+ * Copyright (c) 2026, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.apimgt.gateway.utils;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Manages per-API-context ReentrantLocks to coordinate concurrent API artifact deployments.
+ *
+ * <p>Used to prevent the race condition between two threads that can concurrently attempt to
+ * deploy the same Synapse API artifact when {@code enable_on_demand_loading=true} is configured:
+ * <ul>
+ *   <li><b>Tenant-loading thread</b> (GatewayStartupListener / JMS reconnect) — uses
+ *       {@link #tryLockNow(String)} (non-blocking): if the passthrough thread already holds
+ *       the lock, the startup thread skips deployment because the API will be deployed by
+ *       the passthrough thread anyway.</li>
+ *   <li><b>Passthrough thread</b> (DefaultAPIHandler on-demand path) — uses {@link #lock(String)}
+ *       (blocking): waits until the tenant-loading thread releases the lock before proceeding,
+ *       then re-checks {@code isDeployed()} before actually invoking deploy.</li>
+ * </ul>
+ *
+ * <p>Locks are reference-counted so that map entries are cleaned up automatically when no
+ * thread references a given key.
+ */
+public class APILockManager {
+
+    private static final Log log = LogFactory.getLog(APILockManager.class);
+    private static final APILockManager INSTANCE = new APILockManager();
+
+    /**
+     * Internal lock entry: a fair ReentrantLock paired with a reference counter.
+     */
+    private static class LockEntry {
+        final ReentrantLock lock;
+        final AtomicInteger refCount = new AtomicInteger(0);
+
+        LockEntry() {
+            this.lock = new ReentrantLock(true /* fair */);
+        }
+    }
+
+    private final Map<String, LockEntry> lockMap = new ConcurrentHashMap<>();
+
+    private APILockManager() {
+    }
+
+    public static APILockManager getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     * Acquires the lock for the given key, blocking until it is available.
+     * Intended for the passthrough (on-demand loading) thread.
+     *
+     * @param key unique lock key (typically {@code API_LOADING_ON_DEMAND + apiContext})
+     */
+    public void lock(String key) {
+        LockEntry entry = lockMap.compute(key, (k, v) -> {
+            LockEntry e = (v == null) ? new LockEntry() : v;
+            e.refCount.incrementAndGet();
+            return e;
+        });
+        entry.lock.lock();
+    }
+
+    /**
+     * Attempts to acquire the lock for the given key immediately without waiting.
+     * Intended for the tenant-loading / JMS-redeploy thread.
+     *
+     * @param key unique lock key
+     * @return {@code true} if the lock was acquired; {@code false} if another thread holds it
+     */
+    public boolean tryLockNow(String key) {
+        LockEntry entry = lockMap.compute(key, (k, v) -> {
+            LockEntry e = (v == null) ? new LockEntry() : v;
+            e.refCount.incrementAndGet();
+            return e;
+        });
+        if (!entry.lock.tryLock()) {
+            lockMap.compute(key, (k, v) -> {
+                if (v == null) {
+                    return null;
+                }
+                return v.refCount.decrementAndGet() == 0 ? null : v;
+            });
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Returns {@code true} if the lock for the given key is currently held by any thread, or
+     * if there are threads waiting/referencing it ({@code refCount > 0}).
+     *
+     * @param key unique lock key
+     * @return {@code true} if locked or in use
+     */
+    public boolean isLocked(String key) {
+        LockEntry entry = lockMap.get(key);
+        if (entry == null) {
+            return false;
+        }
+        return entry.lock.isLocked() || entry.refCount.get() > 0;
+    }
+
+    /**
+     * Releases the lock for the given key. Must be called by the thread that holds the lock,
+     * typically in a {@code finally} block.
+     *
+     * @param key unique lock key
+     */
+    public void unlock(String key) {
+        lockMap.compute(key, (k, v) -> {
+            if (v == null) {
+                log.error("APILockManager: attempted to unlock a key with no entry: " + key);
+                return null;
+            }
+            if (!v.lock.isHeldByCurrentThread()) {
+                log.error("APILockManager: current thread does not hold lock for key: " + key);
+                return v;
+            }
+            v.lock.unlock();
+            return v.refCount.decrementAndGet() == 0 ? null : v;
+        });
+    }
+}
+

--- a/components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/handlers/DefaultAPIHandlerTest.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/handlers/DefaultAPIHandlerTest.java
@@ -29,6 +29,7 @@ import org.mockito.Mockito;
 import org.powermock.api.mockito.PowerMockito;
 import org.powermock.core.classloader.annotations.PrepareForTest;
 import org.powermock.modules.junit4.PowerMockRunner;
+import org.wso2.carbon.apimgt.gateway.InMemoryAPIDeployer;
 import org.wso2.carbon.apimgt.gateway.internal.ServiceReferenceHolder;
 import org.wso2.carbon.apimgt.gateway.utils.GatewayUtils;
 import org.wso2.carbon.apimgt.impl.APIConstants;
@@ -41,7 +42,9 @@ import java.util.Set;
 import java.util.TreeMap;
 
 @RunWith(PowerMockRunner.class)
-@PrepareForTest({ApiUtils.class, Utils.class, GatewayUtils.class, ServiceReferenceHolder.class, APIManagerConfiguration.class, ExtendedJWTConfigurationDto.class})
+@PrepareForTest({ApiUtils.class, Utils.class, GatewayUtils.class, ServiceReferenceHolder.class,
+        APIManagerConfiguration.class, ExtendedJWTConfigurationDto.class, InMemoryAPIDeployer.class,
+        DefaultAPIHandler.class})
 public class DefaultAPIHandlerTest {
 
     private ServiceReferenceHolder serviceReferenceHolder;
@@ -131,5 +134,93 @@ public class DefaultAPIHandlerTest {
         Mockito.verify(axis2MsgCntxt, Mockito.times(0)).setProperty(APIConstants.TRANSPORT_URL_IN, "/api1/abc/1.0" +
                 ".0/cde?c=a");
         Mockito.verify(properties, Mockito.times(0)).remove(RESTConstants.REST_FULL_REQUEST_PATH);
+    }
+
+    /**
+     * Consideration 3 — Test (a): On-demand loading is enabled and the API is NOT yet deployed.
+     * Verifies that {@code InMemoryAPIDeployer.deployAPI()} is called exactly once and the handler
+     * returns {@code true}.
+     *
+     * <p>This covers the path where the passthrough thread is the first to detect the undeployed
+     * state and acquires the APILockManager lock before deploying.
+     */
+    @Test
+    public void testOnDemandLoadingDeploysWhenNotDeployed() throws Exception {
+        MessageContext messageContext = Mockito.mock(Axis2MessageContext.class);
+        org.apache.axis2.context.MessageContext axis2MsgCntxt =
+                Mockito.mock(org.apache.axis2.context.MessageContext.class);
+        Mockito.when(((Axis2MessageContext) messageContext).getAxis2MessageContext()).thenReturn(axis2MsgCntxt);
+        Mockito.when(axis2MsgCntxt.getProperty(APIConstants.TRANSPORT_URL_IN)).thenReturn("/api1/abc/cde");
+        PowerMockito.when(GatewayUtils.getTenantDomain()).thenReturn("test.com");
+        PowerMockito.when(GatewayUtils.isOnDemandLoading()).thenReturn(true);
+        PowerMockito.when(ApiUtils.getFullRequestPath(messageContext)).thenReturn("/api1/abc/cde");
+        PowerMockito.when(GatewayUtils.checkForFileBasedApiContexts("/api1/abc/cde", "test.com"))
+                .thenReturn(false);
+
+        // API is NOT deployed initially
+        API api = new API("uuid-1", 1, "admin", "API1", "1.0.0", "/api1/abc/1.0.0", null, "HTTP",
+                "PUBLISHED", false, false);
+        TreeMap<String, API> apiTreeMap = new TreeMap<>();
+        apiTreeMap.put("/api1/abc", api);
+        PowerMockito.when(Utils.getSelectedAPIList("/api1/abc/cde", "test.com")).thenReturn(apiTreeMap);
+
+        // Mock new InMemoryAPIDeployer() so no real deployment is attempted
+        InMemoryAPIDeployer mockDeployer = Mockito.mock(InMemoryAPIDeployer.class);
+        PowerMockito.whenNew(InMemoryAPIDeployer.class).withNoArguments().thenReturn(mockDeployer);
+        Mockito.when(mockDeployer.deployAPI(Mockito.anyString())).thenReturn(true);
+
+        Set<String> properties = Mockito.mock(Set.class);
+        Mockito.when(messageContext.getPropertyKeySet()).thenReturn(properties);
+        Mockito.when(properties.contains(InboundWebsocketConstants.WEBSOCKET_SUBSCRIBER_PATH)).thenReturn(false);
+
+        DefaultAPIHandler handler = new DefaultAPIHandler();
+        boolean result = handler.handleRequestInFlow(messageContext);
+
+        Assert.assertTrue("Handler should return true after on-demand deploy", result);
+        // deployAPI must be invoked once since the API was not deployed
+        Mockito.verify(mockDeployer, Mockito.times(1)).deployAPI("uuid-1");
+    }
+
+    /**
+     * Consideration 3 — Test (b): On-demand loading is enabled but the API is ALREADY deployed
+     * (e.g., tenant-loading thread won the race and set isDeployed=true before the passthrough
+     * thread re-checked after acquiring the APILockManager lock).
+     *
+     * <p>Verifies that {@code InMemoryAPIDeployer.deployAPI()} is NOT called (idempotent
+     * re-entry) and the handler returns {@code true}.
+     */
+    @Test
+    public void testOnDemandLoadingSkipsDeployWhenAlreadyDeployed() throws Exception {
+        MessageContext messageContext = Mockito.mock(Axis2MessageContext.class);
+        org.apache.axis2.context.MessageContext axis2MsgCntxt =
+                Mockito.mock(org.apache.axis2.context.MessageContext.class);
+        Mockito.when(((Axis2MessageContext) messageContext).getAxis2MessageContext()).thenReturn(axis2MsgCntxt);
+        Mockito.when(axis2MsgCntxt.getProperty(APIConstants.TRANSPORT_URL_IN)).thenReturn("/api1/abc/cde");
+        PowerMockito.when(GatewayUtils.getTenantDomain()).thenReturn("test.com");
+        PowerMockito.when(GatewayUtils.isOnDemandLoading()).thenReturn(true);
+        PowerMockito.when(ApiUtils.getFullRequestPath(messageContext)).thenReturn("/api1/abc/cde");
+        PowerMockito.when(GatewayUtils.checkForFileBasedApiContexts("/api1/abc/cde", "test.com"))
+                .thenReturn(false);
+
+        // API is already deployed — simulates tenant-loading thread completing before passthrough
+        API api = new API("uuid-2", 1, "admin", "API1", "1.0.0", "/api1/abc/1.0.0", null, "HTTP",
+                "PUBLISHED", false, true /* isDeployed = true */);
+        TreeMap<String, API> apiTreeMap = new TreeMap<>();
+        apiTreeMap.put("/api1/abc", api);
+        PowerMockito.when(Utils.getSelectedAPIList("/api1/abc/cde", "test.com")).thenReturn(apiTreeMap);
+
+        InMemoryAPIDeployer mockDeployer = Mockito.mock(InMemoryAPIDeployer.class);
+        PowerMockito.whenNew(InMemoryAPIDeployer.class).withNoArguments().thenReturn(mockDeployer);
+
+        Set<String> properties = Mockito.mock(Set.class);
+        Mockito.when(messageContext.getPropertyKeySet()).thenReturn(properties);
+        Mockito.when(properties.contains(InboundWebsocketConstants.WEBSOCKET_SUBSCRIBER_PATH)).thenReturn(false);
+
+        DefaultAPIHandler handler = new DefaultAPIHandler();
+        boolean result = handler.handleRequestInFlow(messageContext);
+
+        Assert.assertTrue("Handler should return true when API is already deployed", result);
+        // deployAPI must NOT be called — the API is already deployed
+        Mockito.verify(mockDeployer, Mockito.never()).deployAPI(Mockito.anyString());
     }
 }

--- a/components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/utils/APILockManagerTest.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.gateway/src/test/java/org/wso2/carbon/apimgt/gateway/utils/APILockManagerTest.java
@@ -1,0 +1,242 @@
+/*
+ * Copyright (c) 2026, WSO2 LLC. (http://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.apimgt.gateway.utils;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Unit tests for {@link APILockManager}.
+ *
+ * <p>Covers:
+ * <ul>
+ *   <li>Scenario 1 — Tenant-loading thread wins (tryLockNow first), passthrough thread blocks on
+ *       lock() and re-checks isDeployed() after the startup thread releases.</li>
+ *   <li>Scenario 2 — Passthrough thread wins (lock() first), tryLockNow in startup thread returns
+ *       false → startup thread skips deployment.</li>
+ *   <li>Scenario 3 — isLocked() reflects real state before and after unlock.</li>
+ *   <li>Scenario 4 — Lock entry is cleaned up from the internal map after unlock, preventing
+ *       memory leaks over time.</li>
+ *   <li>Scenario 5 — tryLockNow is idempotent for the same key (second call returns false while
+ *       first hold is still active).</li>
+ *   <li>Scenario 6 — Concurrent mutual exclusion: N threads all try to deploy; exactly one
+ *       increments the counter per wave.</li>
+ * </ul>
+ */
+public class APILockManagerTest {
+
+    private final APILockManager lockManager = APILockManager.getInstance();
+    private static final String TEST_KEY = "LoadAPI_OnDemandTest_/t/test.com/pizzashack/1.0.0";
+
+    /**
+     * Release the lock only if it is still held at the end of a test (e.g., a mid-test failure),
+     * so passing tests do not exercise the error-log path inside unlock().
+     */
+    @After
+    public void tearDown() {
+        if (lockManager.isLocked(TEST_KEY)) {
+            lockManager.unlock(TEST_KEY);
+        }
+    }
+
+    /**
+     * Scenario 1: Tenant-loading thread (tryLockNow) acquires the lock first.
+     * Passthrough thread on lock() must block, then re-check after the lock is released.
+     */
+    @Test
+    public void testStartupThreadWinsPassthroughBlocks() throws InterruptedException {
+        // Simulate startup thread acquiring the lock non-blocking
+        boolean startupAcquired = lockManager.tryLockNow(TEST_KEY);
+        Assert.assertTrue("Startup thread should acquire an uncontested lock", startupAcquired);
+        Assert.assertTrue("Lock should be reported as locked", lockManager.isLocked(TEST_KEY));
+
+        AtomicBoolean passthroughLockAcquired = new AtomicBoolean(false);
+        CountDownLatch passthroughStarted = new CountDownLatch(1);
+        CountDownLatch passthroughDone = new CountDownLatch(1);
+
+        // Simulate passthrough thread attempting to acquire (should block)
+        Thread passthroughThread = new Thread(() -> {
+            passthroughStarted.countDown(); // signal that it's about to block
+            lockManager.lock(TEST_KEY);
+            try {
+                passthroughLockAcquired.set(true);
+            } finally {
+                lockManager.unlock(TEST_KEY);
+                passthroughDone.countDown();
+            }
+        });
+        passthroughThread.start();
+
+        // Wait until the passthrough thread has started
+        Assert.assertTrue(passthroughStarted.await(2, TimeUnit.SECONDS));
+
+        // Wait deterministically until the passthrough thread is parked inside lock()
+        // (Thread.State.WAITING == LockSupport.park, which ReentrantLock.lock() uses).
+        // This replaces a fixed Thread.sleep() that is unreliable on slow/loaded CI runners.
+        long deadline = System.currentTimeMillis() + 2000;
+        while (passthroughThread.getState() != Thread.State.WAITING
+                && System.currentTimeMillis() < deadline) {
+            Thread.yield();
+        }
+        Assert.assertEquals("Passthrough thread should be parked inside lock()",
+                Thread.State.WAITING, passthroughThread.getState());
+
+        // Passthrough thread is definitively blocked — it must NOT have acquired the lock yet
+        Assert.assertFalse("Passthrough should still be blocked while startup holds lock",
+                passthroughLockAcquired.get());
+
+        // Startup thread finishes — release the lock
+        lockManager.unlock(TEST_KEY);
+
+        // Now passthrough thread should unblock and acquire
+        Assert.assertTrue("Passthrough thread should unblock after startup releases",
+                passthroughDone.await(3, TimeUnit.SECONDS));
+        Assert.assertTrue("Passthrough thread should have acquired the lock after startup released",
+                passthroughLockAcquired.get());
+    }
+
+    /**
+     * Scenario 2: Passthrough thread holds the blocking lock.
+     * Startup thread's tryLockNow must return false (non-blocking skip behaviour).
+     */
+    @Test
+    public void testPassthroughThreadWinsStartupSkips() throws InterruptedException {
+        CountDownLatch passthroughHoldsLock = new CountDownLatch(1);
+        CountDownLatch startupChecked = new CountDownLatch(1);
+
+        Thread passthroughThread = new Thread(() -> {
+            lockManager.lock(TEST_KEY);
+            try {
+                passthroughHoldsLock.countDown(); // signal: lock is held
+                // Wait until startup thread has done its tryLockNow check
+                try {
+                    startupChecked.await(3, TimeUnit.SECONDS);
+                } catch (InterruptedException ignored) {
+                }
+            } finally {
+                lockManager.unlock(TEST_KEY);
+            }
+        });
+        passthroughThread.start();
+
+        // Wait until passthrough holds the lock
+        Assert.assertTrue(passthroughHoldsLock.await(2, TimeUnit.SECONDS));
+
+        // Startup thread's tryLockNow should return false — API is being deployed by passthrough
+        boolean startupAcquired = lockManager.tryLockNow(TEST_KEY);
+        startupChecked.countDown(); // signal done
+
+        Assert.assertFalse("Startup thread should NOT acquire lock when passthrough holds it",
+                startupAcquired);
+        passthroughThread.join(3000);
+    }
+
+    /**
+     * Scenario 3: isLocked() correctly reflects lock state transitions.
+     */
+    @Test
+    public void testIsLockedReflectsState() {
+        Assert.assertFalse("Key should not be locked before any lock call",
+                lockManager.isLocked(TEST_KEY));
+
+        lockManager.lock(TEST_KEY);
+        Assert.assertTrue("Key should be reported locked after lock()", lockManager.isLocked(TEST_KEY));
+
+        lockManager.unlock(TEST_KEY);
+        Assert.assertFalse("Key should not be locked after unlock()", lockManager.isLocked(TEST_KEY));
+    }
+
+    /**
+     * Scenario 4: Lock entry is removed from internal map after all references are released,
+     * preventing memory leaks over time.
+     */
+    @Test
+    public void testLockEntryCleanedUpAfterUnlock() {
+        lockManager.lock(TEST_KEY);
+        lockManager.unlock(TEST_KEY);
+
+        // After unlock with no other references, isLocked should return false (entry removed)
+        Assert.assertFalse("Lock entry should be cleaned up after unlock", lockManager.isLocked(TEST_KEY));
+    }
+
+    /**
+     * Scenario 5: Second tryLockNow on a key already held returns false (not re-entrant across
+     * different "threads" — a different thread cannot steal the lock).
+     */
+    @Test
+    public void testTryLockNowIsNotReentrantAcrossThreads() throws InterruptedException {
+        // First caller acquires successfully
+        boolean first = lockManager.tryLockNow(TEST_KEY);
+        Assert.assertTrue("First tryLockNow should succeed", first);
+
+        // A different thread tries the same key — must fail
+        AtomicBoolean secondResult = new AtomicBoolean(true);
+        Thread competitor = new Thread(() -> secondResult.set(lockManager.tryLockNow(TEST_KEY)));
+        competitor.start();
+        competitor.join(2000);
+
+        Assert.assertFalse("Second tryLockNow from a different thread should fail",
+                secondResult.get());
+
+        lockManager.unlock(TEST_KEY);
+    }
+
+    /**
+     * Scenario 6 — Concurrent stress: N threads all try to deploy the same API (all calling
+     * lock() then checking a shared counter). Exactly one should increment the counter on each
+     * "wave" without duplication — verifying mutual exclusion.
+     */
+    @Test
+    public void testMutualExclusionUnderConcurrency() throws InterruptedException {
+        final int threadCount = 20;
+        final AtomicInteger deployCount = new AtomicInteger(0);
+        final CountDownLatch startGate = new CountDownLatch(1);
+        final CountDownLatch done = new CountDownLatch(threadCount);
+
+        for (int i = 0; i < threadCount; i++) {
+            new Thread(() -> {
+                try {
+                    startGate.await();
+                } catch (InterruptedException ignored) {
+                }
+                lockManager.lock(TEST_KEY);
+                try {
+                    // Simulate check-then-deploy: only deploy if not yet deployed
+                    if (deployCount.get() == 0) {
+                        deployCount.incrementAndGet();
+                    }
+                } finally {
+                    lockManager.unlock(TEST_KEY);
+                    done.countDown();
+                }
+            }).start();
+        }
+
+        startGate.countDown(); // release all threads simultaneously
+        Assert.assertTrue("All threads should finish", done.await(10, TimeUnit.SECONDS));
+        Assert.assertEquals("Exactly one deployment should have occurred", 1, deployCount.get());
+    }
+}
+

--- a/components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/entity/API.java
+++ b/components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/model/entity/API.java
@@ -43,7 +43,7 @@ public class API implements CacheableEntity<String> {
     private String apiType = null;
     private String status;
     private String organization;
-    private boolean deployed = false;
+    private volatile boolean deployed = false;
     private boolean isDefaultVersion = false;
     private String securityScheme;
     private String revisionId;

